### PR TITLE
Feat video urls

### PR DESCRIPTION
**Pull request by lsh-0**

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -183,6 +184,66 @@ def to_volume(volume):
...
+    sources = {
+        'mp4': 'video/mp4; codecs="avc1.42E01E, mp4a.40.2"',
+        'webm': 'video/webm; codecs="vp8.0, vorbis"',
+        'ogv': 'video/ogg; codecs="theora, vorbis"',
+    }
+
+    def fn(video):
+        try:
+            vid = video['id']
+            ensure(vid in gc_data, "glencoe doesn't know %r, only %s" % (vid, gc_id_str))
```
Comment (giorgiosironi) [inline src/main.py]: in the case of the testing environment, `msid` will be generated (something like `1234567890`) and `gc_data` should be `{}` as there are no video indexed with these generated keys on Glencoe. Will this part explode? Is there a way to make it skip not found videos (or should we try to load them with the original id?)

Comment (lsh-0) [inline src/main.py]: thats what the try ... catch is for. the kitchen sink doesn't have glencoe videos either.

Comment (lsh-0) [inline src/main.py]: ah, is this: `15:24:02 INFO - 2016-11-30 15:18:09,353 - elife-09977-v2.xml -> elife-09977-v2.xml.json => failed ('content')` what you referring to?

**Comment by lsh-0** [comment]
```
15:18:43 Traceback (most recent call last):
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 399, in main
15:18:43     article_json = render_single(doc, version=version)
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 391, in render_single
15:18:43     return postprocess(render(description, [soup])[0])
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 273, in postprocess
15:18:43     data['article'] = doall(article, [expand_any_videos, clean_if_none, clean_if_empty])
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/venv/local/lib/python2.7/site-packages/et3/render.py", line 12, in doall
15:18:43     item = seg(item)
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 244, in expand_any_videos
15:18:43     article['body'] = do_body_for_type(body_content, ['video'], fn)
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 208, in do_body_for_type
15:18:43     return do_body(body_content, _fn)
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 203, in do_body
15:18:43     return map(lambda element: do_section(element, fn), body_content)
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 203, in <lambda>
15:18:43     return map(lambda element: do_section(element, fn), body_content)
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 196, in do_section
15:18:43     element['content'] = [do_section(child, fn) for child in element['content']]
15:18:43   File "/ext/jenkins-libraries-runner/workspace/7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q/src/main.py", line 196, in do_section
15:18:43     element['content'] = [do_section(child, fn) for child in element['content']]
15:18:43 KeyError: 'content'
```

File: src/main.py (hunk 1 above)
Comment (lsh-0) [inline src/main.py]: indeed. the current failure is to do with a `section` not having `content`.

**Comment by lsh-0** [comment]
```
18:41:35 
18:41:35 real	4m1.153s
18:41:35 user	31m53.374s
18:41:35 sys	0m4.704s
[Pipeline] archive
[Pipeline] sh
18:41:35 [7-BMR5WBONHLWAVZUMETR34CCQZ275ICCMAWJVJWGKRAEQMRYG3N6Q] Running shell script
18:41:36 + ./validate-statistics.sh validation.log
18:41:36 Input: 4912 JSON articles
18:41:36 Output valid: 4822 JSON articles
18:41:36 Output invalid: 90 JSON articles
18:41:36 Validated green,red,total: 4822,90,4912
18:41:36 No more than 88 articles should be failing validation (with placeholders)
```

bah! bullshit. I give up for today. fml.

**Comment by gnott** [comment]
I was curious why the invalid count increased by 2. Compared the last valid run it looks like these two are now invalid and were not invalid before

elife-07574-v2.xml.json
elife-13245-v2.xml.json

**Comment by gnott** [comment]
1. On 07574, I think it is invalid on Video 2. u'sources': []

https://movie-usa.glencoesoftware.com/metadata/10.7554/eLife.07574

There's some bad data I think. "elife-07574-media2.mp4" is associated with "10.7554/eLife.07574.007" in the metadata, but in the XML it says it is 10.7554/eLife.07574.012.

10.7554/eLife.07574.012 is associated with a file called "elife07574v001." which will not exist.

This one in particular probably needs to be resupplied to Glencoe and should be reviewed by Production.

2. https://movie-usa.glencoesoftware.com/metadata/10.7554/eLife.13245  this one seems to have no video metadata. Will also need review and possibly resupply to Glencoe.

**Comment by gnott** [comment]
http://jira.elifesciences.org:8080/browse/PI-214

**Comment by giorgiosironi** [comment]
What was the problem about 900MB of output in the build by the way?

**Comment by lsh-0** [comment]
@giorgiosironi that was every article with videos failing validation because of http vs https

**Comment by lsh-0** [comment]
@gnott ` u'sources': []` - I left an empty array in there for articles that don't have a mention in glencoe. like kitchen sink and test fixtures. there is no mention of min size in the api raml, I thought an empty list would be fine.

**Pull request merged**